### PR TITLE
adding `callback_on_step_end` for `StableDiffusionLDM3DPipeline`

### LINKED ISSUE #7136: adding `callback_on_step_end` for `StableDiffusionLDM3DPipeline`
Just realize the StableDiffusionLDM3DPipeline is still not using the new callback API
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py

Is anyone interested in refactoring it? you can use this PR as a reference  
https://github.com/huggingface/diffusers/pull/5812

### PR DESCRIPTION
**Pull request by rootonchair**
# What does this PR do?

Fixes #7136 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by rootonchair** [comment]
Test script
```python
import torch
from diffusers import StableDiffusionLDM3DPipeline

def callback_dynamic_cfg(pipe, step_index, timestep, callback_kwargs):
    # adjust the batch_size of prompt_embeds according to guidance_scale
    if step_index == int(pipe.num_timesteps * 0.4):
        prompt_embeds = callback_kwargs["prompt_embeds"]
        prompt_embeds = prompt_embeds.chunk(2)[-1]

        # update guidance_scale and prompt_embeds
        pipe._guidance_scale = 0.0
        callback_kwargs["prompt_embeds"] = prompt_embeds
    return callback_kwargs

pipe = StableDiffusionLDM3DPipeline.from_pretrained("Intel/ldm3d-4c")
# pipe = pipe.to("cuda")

prompt = "a photo of an astronaut riding a horse on mars"
generator = torch.Generator()
generator.manual_seed(42)
output = pipe(prompt, generator=generator, callback_on_step_end=callback_dynamic_cfg, callback_on_step_end_tensor_inputs=['prompt_embeds'])
rgb_image, depth_image = output.rgb, output.depth
rgb_image[0].save("astronaut_ldm3d_rgb.jpg")
depth_image[0].save("astronaut_ldm3d_depth.png")
```

File: src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py [hunk 1]
```diff
@@ -59,6 +59,64 @@
 """
 
 
+def rescale_noise_cfg(noise_cfg, noise_pred_text, guidance_rescale=0.0):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py:62]: add #copied from?

File: src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py [hunk 2]
```diff
@@ -59,6 +59,64 @@
...
+    std_text = noise_pred_text.std(dim=list(range(1, noise_pred_text.ndim)), keepdim=True)
+    std_cfg = noise_cfg.std(dim=list(range(1, noise_cfg.ndim)), keepdim=True)
+    # rescale the results from guidance (fixes overexposure)
+    noise_pred_rescaled = noise_cfg * (std_text / std_cfg)
+    # mix with the original results from guidance by factor guidance_rescale to avoid "plain looking" images
+    noise_cfg = guidance_rescale * noise_pred_rescaled + (1 - guidance_rescale) * noise_cfg
+    return noise_cfg
+
+
+def retrieve_timesteps(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py:76]: add a `#Copied from`

**Review by yiyixuxu — COMMENTED** [review]
looking great! thanks!
think we are just missing some `#copied` from

File: src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py (hunk 2 above)
Comment (rootonchair) [reply src/diffusers/pipelines/stable_diffusion_ldm3d/pipeline_stable_diffusion_ldm3d.py:78]: Sure, I will

**Review by yiyixuxu — APPROVED** [review]

**Pull request merged**